cpp/src/strings/convert/convert_urls.cu

```diff
@@ -325,85 +378,60 @@ std::unique_ptr<column> url_decode(
 
   // build offsets column
   auto offsets_column = make_numeric_column(
     data_type{type_id::INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
-  auto offsets_view = offsets_column->mutable_view();
-  thrust::transform(rmm::exec_policy(stream),
-                    d_offsets_span.begin(),
-                    d_offsets_span.end(),
-                    offsets_view.begin<int32_t>(),
-                    url_decode_offsets_updater{d_esc_positions_span, chars_start});
+
+  // count number of bytes in each string after decoding and store it in offsets_column
+  auto offsets_view         = offsets_column->view();
+  auto offsets_mutable_view = offsets_column->mutable_view();
+  url_decode_char_counter<threadblock_size, block_size>
+    <<<num_threadblocks, threadblock_size, 0, stream.value()>>>(
+      d_in_chars, d_in_offsets, offsets_mutable_view.begin<int32_t>() + 1, strings_count);
+
+  // use inclusive_scan to transform number of bytes into offsets
+  thrust::inclusive_scan(rmm::exec_policy(stream),
+                         offsets_view.begin<int32_t>() + 1,
+                         offsets_view.begin<int32_t>() + offset_count,
+                         offsets_mutable_view.begin<int32_t>() + 1);
+
+  // set the first element of the offset column to 0
+  CUDA_TRY(
+    cudaMemsetAsync(offsets_mutable_view.begin<int32_t>(), 0, sizeof(int32_t), stream.value()));
+
+  // copy the total number of characters of all strings combined (last element of the offset column)
+  // to the host memory
+  int32_t out_chars_bytes;
+  CUDA_TRY(cudaMemcpyAsync(&out_chars_bytes,
+                           offsets_view.begin<int32_t>() + offset_count - 1,
+                           sizeof(int32_t),
+                           cudaMemcpyDeviceToHost,
+                           stream.value()));
+  stream.synchronize();
 
   // create the chars column
-  auto chars_column =
-    create_chars_child_column(chars_bytes - (esc_count * 2),  // replacing 3 bytes with 1
-                              stream,
-                              mr);
-  auto d_out_chars = chars_column->mutable_view().data<char>();
-  thrust::for_each_n(
-    rmm::exec_policy(stream),
-    thrust::make_counting_iterator<size_type>(chars_start),
-    chars_bytes,
-    url_decode_char_replacer{d_esc_positions_span, d_in_chars, d_out_chars, chars_start});
-
-  // free the escape positions buffer as it is no longer needed
-  (void)esc_positions.release();
+  auto chars_column = create_chars_child_column(out_chars_bytes, stream, mr);
+  auto d_out_chars  = chars_column->mutable_view().data<char>();
+
+  // decode and copy the characters from the input column to the output column
+  url_decode_char_replacer<threadblock_size, block_size>
+    <<<num_threadblocks, threadblock_size, 0, stream.value()>>>(
+      d_in_chars,
+      d_in_offsets,
+      d_out_chars,
+      offsets_column->view().begin<int32_t>(),
+      strings_count);
```

Comment:
The entire of this function behaves very similarly to the existing string APIs: A 2 steps approach in which the first step computes the output strings sizes and offsets, and the second step computes the output strings. Let use `cudf::strings::detail::make_strings_children` with functor. You can find examples of its usage in many places in cudf.

Reply:
Thanks for the quick review and the suggestion!

I have some doubts about the performance of `cudf::strings::detail::make_strings_children`. Looking at its implementation in [cudf/strings/detail/utilities.cuh](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/strings/detail/utilities.cuh), it assigns each string to a thread through `thrust::for_each_n`. I think this could lead to poor performance for long strings because the load/store is not coalesced. This is supported by the performance numbers reported in https://github.com/rapidsai/cudf/pull/7353, when we switch from string-level parallelism to char-level parallelism for performance's reasons.

Please correct me if my understanding of `cudf::strings::detail::make_strings_children` is not correct.

Reply:
That API has added a parameter `exec_size` to allow having the number of strings different from the number of execution threads. You can specify the number of threads to be the number of characters.

Reply:
Thanks for the clarification!  Just to be clear, I am not looking for assigning characters to threads. There are three strategies for the work assignment here:
1. assign strings to threads (implementation before https://github.com/rapidsai/cudf/pull/7353).
2. assign characters to threads (implementation of https://github.com/rapidsai/cudf/pull/7353).
3. assign strings to warps (this PR).

As discussed in the previous comment, the problem with strategy 1 is that memory accesses are not coalesced. The problem with strategy 2 is that we need binary searches in global memory for the correct output locations as well as detecting string boundaries, and these binary searches are costly. Strategy 3 improves on strategy 2 in two areas:
- Since we assigns strings to warps, we can read the string index from the warp index, and the string boundary is known.
- Threads in a warp could use warp-level reduction to compute the output location.

The result is that we can get rid of the binary searches altogether.

With `make_strings_children`, I guess we could simulate assigning strings to warps by setting `exec_size` to `32 * num of strings`, and assign strings to every 32 threads in a round-robin fashion. But could we still use shared memory and warp-level reduction? I think that's probably not a good idea because we cannot guarantee how thrust maps from elements to threads.

Reply:
I agree that you should not be using `make_strings_children` for this optimization.